dbms/src/Flash/Coprocessor/DAGExpressionAnalyzer.cpp

```diff
@@ -192,6 +200,153 @@ void appendWindowDescription(
+{
+    // Execute this function only when the frame type is Range
+    if (window.frame().type() != tipb::WindowFrameType::Ranges)
+        return std::make_pair("", "");
+
+    String begin_aux_col_name;
+    String end_aux_col_name;
+
+    // For begin frame
+    if (window.frame().start().has_frame_range())
+        begin_aux_col_name = DAGExpressionAnalyzerHelper::buildFunction(analyzer, window.frame().start().frame_range(), actions);
+    // For end frame
+    if (window.frame().end().has_frame_range())
+        end_aux_col_name = DAGExpressionAnalyzerHelper::buildFunction(analyzer, window.frame().end().frame_range(), actions);
+
+    return std::make_pair(begin_aux_col_name, end_aux_col_name);
+}
+
+WindowDescription createAndInitWindowDesc(DAGExpressionAnalyzer * const analyzer, const tipb::Window & window)
+{
+    WindowDescription window_description;
+    window_description.partition_by = analyzer->getWindowSortDescription(window.partition_by());
+    window_description.order_by = analyzer->getWindowSortDescription(window.order_by());
+    if (window.has_frame())
+    {
+        window_description.setWindowFrame(window.frame());
+    }
+
+    return window_description;
+}
+
+void initBeforeWindow(
+    DAGExpressionAnalyzer * analyzer,
+    WindowDescription & window_desc,
+    ExpressionActionsChain & chain,
+    const tipb::Window & window,
+    ExpressionActionsChain::Step & step)
+{
+    // Prepare auxiliary function for range frame type
+    auto aux_col_names = addRangeFrameAuxiliaryFunctionAction(analyzer, step.actions, window);
+
+    analyzer->appendWindowColumns(window_desc, window, step.actions);
+    // set required output for window funcs's arguments.
+    for (const auto & window_function_description : window_desc.window_functions_descriptions)
+    {
+        for (const auto & argument_name : window_function_description.argument_names)
+            step.required_output.push_back(argument_name);
+    }
+
+    window_desc.before_window = chain.getLastActions();
+    setAuxiliaryColumnInfo(step.actions, window_desc, aux_col_names.first, aux_col_names.second, window);
+
+    // We should not call finalize here as some useful columns will be unexpected removed
+    chain.clear();
+}
+
+void initAfterWindow(
+    DAGExpressionAnalyzer * const analyze,
+    WindowDescription & window_desc,
+    ExpressionActionsChain & chain,
+    const tipb::Window & window,
+    size_t source_size)
+{
+    auto & after_window_step = analyze->initAndGetLastStep(chain);
+    analyze->appendCastAfterWindow(after_window_step.actions, window, source_size);
+    window_desc.after_window_columns = analyze->getCurrentInputColumns();
+    analyze->appendSourceColumnsToRequireOutput(after_window_step);
+    window_desc.after_window = chain.getLastActions();
+    chain.finalize();
+    chain.clear();
+}
+
+ExpressionActionsChain::Step & createStepForBuildingWindowDescription(DAGExpressionAnalyzer * const analyzer, ExpressionActionsChain & chain)
```

Comment:
> The function name is a bit strange, because there is no logic related to window here. Is it necessary to wrap a new function here?

I will change this function's name.